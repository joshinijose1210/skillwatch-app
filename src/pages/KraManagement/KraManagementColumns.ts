import { TableColumnConfig } from '@medly-components/core';

export const KraManagementColumns: TableColumnConfig[] = [
    {
        title: 'KRA ID',
        field: 'kraId'
    },
    {
        title: 'KRA',
        field: 'name'
    },
    {
        title: 'Weightage(%)',
        field: 'weightage',
        align: 'center'
    }
];
