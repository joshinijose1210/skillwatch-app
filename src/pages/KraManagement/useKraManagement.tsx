import { routeConstants } from '@constants';
import { useAppSelector } from '@slice';
import { useGetKrasQuery, useGetReviewCycleDataMutation } from '@slice/services';
import { useEffect, useState } from 'react';
import { useNavigate, useParams } from 'react-router-dom';

export interface KraType {
    id: number;
    kraId: string;
    name: string;
    weightage: number;
    organisationId: number;
}

export const useKraManagement = () => {
    const navigate = useNavigate();
    const userDetails = useAppSelector(state => state.user);
    const { ActivePage } = useParams();
    const { modulePermission } = useAppSelector(state => state.user);
    const module = modulePermission?.find(module => module.moduleName === 'KRAs');
    const [page, setPage] = useState<number>();
    const [kras, setKras] = useState<KraType[] | null>(null);
    const [enableEditKra, setEnableEditKra] = useState(false);

    const {
        data,
        isSuccess,
        isFetching: isLoading,
        error
    } = useGetKrasQuery(
        {
            path: '',
            params: [
                { name: 'page', value: ActivePage ?? page },
                { name: 'limit', value: 10 },
                { name: 'organisationId', value: userDetails.organisationId }
            ]
        },
        { refetchOnMountOrArgChange: true }
    );
    const [getActiveReviewCycle, { data: activeReviewCycleData, isSuccess: hasFetchedActiveCycle }] = useGetReviewCycleDataMutation();

    useEffect(() => {
        if (data && isSuccess) {
            setKras(data);
        }
    }, [data, isSuccess]);

    const openAddNewKRA = () => {
        navigate(`${routeConstants.kraManagement}/1/edit-KRA`);
    };

    useEffect(() => {
        getActiveReviewCycle({
            path: 'active',
            params: [{ name: 'organisationId', value: userDetails.organisationId }]
        });
    }, [getActiveReviewCycle, userDetails.organisationId]);

    // not using useMemo here because we button is disabled by default until after API call completes and conditions are evaluated.
    useEffect(() => {
        if (
            activeReviewCycleData &&
            new Date(activeReviewCycleData?.selfReviewStartDate) > new Date() &&
            new Date(activeReviewCycleData?.checkInWithManagerEndDate) > new Date()
        ) {
            setEnableEditKra(true);
        }
        if (!activeReviewCycleData?.selfReviewStartDate && hasFetchedActiveCycle) {
            setEnableEditKra(true);
        }
    }, [activeReviewCycleData, hasFetchedActiveCycle]);

    return { openAddNewKRA, kras, error, isLoading, page, module, enableEditKra };
};
