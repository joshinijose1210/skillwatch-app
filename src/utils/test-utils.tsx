import { ToastContainer } from '@medly-components/core';
import { setupListeners } from '@reduxjs/toolkit/dist/query';
import { store as rtkStore, setupStore } from '@slice';
import { RenderOptions, RenderResult, render } from '@testing-library/react';
import { defaultTheme } from '@theme';
import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import reduxMockStore from 'redux-mock-store';
import { ThemeProvider } from 'styled-components';
import { userMockState } from '../mocks/userMockState';
import { initialState } from '../slice';

interface CustomOptions extends RenderOptions {
    initialState?: any;
    initialEntries?: any;
}

interface WithThemeProviderProps {
    children?: React.ReactNode;
}

const mockStore = reduxMockStore(),
    store = mockStore({
        ...initialState,
        user: { ...initialState.user, groups: ['admin'] }
    });

const WithThemeProvider: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
    <ThemeProvider theme={defaultTheme}>
        <>
            <ToastContainer position="top-end" />
            {children}
        </>
    </ThemeProvider>
);

const WithStore: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
    <Provider store={store}>
        <WithThemeProvider>{children}</WithThemeProvider>
    </Provider>
);

const WithRouter: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
    <MemoryRouter>
        <WithThemeProvider>{children}</WithThemeProvider>
    </MemoryRouter>
);

const WithStoreAndRouter: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
    <Provider store={store}>
        <MemoryRouter>
            <WithThemeProvider>{children}</WithThemeProvider>
        </MemoryRouter>
    </Provider>
);

export const RTKRender = (ui: React.ReactElement, { initialState = userMockState, ...renderOptions }: CustomOptions = {}) => {
    const store = setupStore(initialState);
    setupListeners(store.dispatch);

    const RtkWrapper: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
        <Provider store={store}>
            <MemoryRouter initialEntries={renderOptions.initialEntries}>
                <WithThemeProvider>{children}</WithThemeProvider>
            </MemoryRouter>
        </Provider>
    );
    return { rtkStore, ...render(ui, { wrapper: RtkWrapper, ...renderOptions }) };
};
const customRender = (ui: React.ReactElement, options?: RenderOptions): RenderResult =>
    render(ui, { wrapper: WithThemeProvider, ...options });

export const renderWithStore = (ui: React.ReactElement, options?: RenderOptions): RenderResult =>
    render(ui, { wrapper: WithStore, ...options });

export const renderWithRouter = (ui: React.ReactElement, options?: RenderOptions): RenderResult =>
    render(ui, { wrapper: WithRouter, ...options });

export const renderWithStoreAndRouter = (ui: React.ReactElement, options?: RenderOptions): RenderResult =>
    render(ui, { wrapper: WithStoreAndRouter, ...options });

export const wrapper = (props: { children: React.ReactNode }, { initialState = userMockState }: CustomOptions = {}) => {
    const store = setupStore(initialState);
    setupListeners(store.dispatch);
    return (
        <Provider store={store}>
            <MemoryRouter>
                <WithThemeProvider>{props.children}</WithThemeProvider>
            </MemoryRouter>
        </Provider>
    );
};

// re-export everything
export * from '@testing-library/react';
// override render method
export { customRender as render, store };
