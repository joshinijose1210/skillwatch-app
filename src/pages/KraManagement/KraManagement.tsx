import { PageContent } from '@components/layout/PageContent/PageContent.styled';
import { CustomTable } from '@components/reusableComponents/CustomTable/CustomTable';
import ListHeader from '@components/reusableComponents/ListHeader';
import { FC } from 'react';
import { KraManagementColumns } from './KraManagementColumns';
import { useKraManagement } from './useKraManagement';
import { ListHeaderButton } from '@components/reusableComponents/ListHeader/ListHeader.styled';
export const KraManagement: FC = () => {
    const { openAddNewKRA, kras, error, module, isLoading, enableEditKra } = useKraManagement();
    return (
        <PageContent>
            {module?.edit ? (
                <ListHeader
                    title="KRAs (Key Responsibility Areas)"
                    moduleTitle="KRA"
                    rightSection={
                        <div>
                            <ListHeaderButton
                                variant="solid"
                                title={!enableEditKra ? 'You cannot edit KRA when self review start date has passed.' : ''}
                                onClick={openAddNewKRA}
                                disabled={!enableEditKra}
                            >
                                Edit KRA
                            </ListHeaderButton>
                        </div>
                    }
                />
            ) : (
                <ListHeader title="KRAs (Key Responsibility Areas)" />
            )}
            <CustomTable
                columns={KraManagementColumns}
                data={kras || []}
                count={0}
                activePage={1}
                isLoading={isLoading}
                handlePageChange={() => null}
            />
        </PageContent>
    );
};
