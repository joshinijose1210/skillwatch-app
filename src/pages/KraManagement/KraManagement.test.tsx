import { cleanup, fireEvent, RTKRender, screen, waitFor } from '@utils/test-utils';
import { kraMockData, permissionModuleWithoutKRA } from '@mocks/kraManagement ';
import { apiInstance } from '@utils';
import { KraManagement } from './KraManagement';
import MockAdapter from 'axios-mock-adapter';
import { routeConstants } from '@constants';

const mock = new MockAdapter(apiInstance, { onNoMatch: 'throwException' });
const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate
}));

beforeAll(() => {
    mock.onGet(kraMockData.kra.get.url).reply(200, kraMockData.kra.get.allData);
    mock.onGet(kraMockData.module.url).reply(200, kraMockData.module.data);
});

beforeEach(mock.reset);

afterEach(cleanup);

describe('KraManagement', () => {
    it('test snapshot', () => {
        const { container } = RTKRender(<KraManagement />);
        expect(container).toMatchSnapshot();
    });

    it('should redirect to edit kra page when click on edit', async () => {
        RTKRender(<KraManagement />);
        const editKraBtn = await screen.findByText('Edit KRA');
        expect(editKraBtn).toBeInTheDocument();
        fireEvent.click(editKraBtn);
        expect(mockNavigate).toHaveBeenCalledWith(`${routeConstants.kraManagement}/1/edit-KRA`);
    });

    it('should not render Edit Kra button if not permitted', async () => {
        RTKRender(<KraManagement />, { initialState: permissionModuleWithoutKRA });
        await waitFor(() => expect(screen.queryByText('Edit KRA')).not.toBeInTheDocument());
    });
});
