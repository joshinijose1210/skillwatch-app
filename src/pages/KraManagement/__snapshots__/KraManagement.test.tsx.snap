// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KraManagement test snapshot 1`] = `
<div>
  <main
    class="sc-1ehavxz-0 dDyhBJ"
  >
    <div
      class="sc-h70dbj-0 fQmYvl row expanded space-between align-center marginBottom-2"
    >
      <div
        class="sc-w8e7z6-2"
      >
        <h3
          class="sc-1ntgl3l-0 kcLBOi sc-w8e7z6-0 ibnmt"
          textalign="initial"
          textcolor="inherit"
          textvariant="h3"
          textweight="Medium"
        >
          KRAs (Key Responsibility Areas)
        </h3>
      </div>
      <div>
        <button
          class="sc-1k2druj-0 eLgfbc sc-w8e7z6-1"
          disabled=""
          edges="square"
          title="You cannot edit KRA when self review start date has passed."
          type="button"
          variant="solid"
        >
          <span
            class="sc-1ntgl3l-0 eOsHHD"
            textalign="initial"
            textcolor="inherit"
            textvariant="button1"
          >
            Edit KRA
          </span>
        </button>
      </div>
    </div>
    <div
      class="sc-8tbhpn-3"
    >
      <div
        class="sc-1msmxxr-0 iWEzjW"
      />
      <table
        actions=""
        class="sc-1msmxxr-1 jwFsPF sc-8tbhpn-2 kCDmJI"
        columns="[object Object],[object Object],[object Object]"
        defaultactivepage="1"
        defaultsortorder="desc"
        itemsperpage="10"
        rowclickdisablekey=""
        tabindex="-1"
        totalitems="0"
      >
        <thead
          class="sc-uxzsxp-0 haNltq"
        >
          <tr
            class="sc-1jr91bg-0 hJzMcO"
            gridtemplatecolumns=" minmax(100px, 1fr) minmax(100px, 1fr) minmax(100px, 1fr)"
            style="grid-template-columns: minmax(100px, 1fr) minmax(100px, 1fr) minmax(100px, 1fr);"
          >
            <th
              class="sc-hjmm4m-0 iOyNiA"
              tablesize="M"
            >
              <button
                class="sc-hjmm4m-2 kpjakb"
                tablesize="M"
              >
                <span
                  class="sc-1ntgl3l-0 ghpJwn"
                  textalign="initial"
                  textcolor="inherit"
                >
                  KRA ID
                </span>
              </button>
              <span
                class="sc-hjmm4m-1 cuEDpa"
              />
            </th>
            <th
              class="sc-hjmm4m-0 iOyNiA"
              tablesize="M"
            >
              <button
                class="sc-hjmm4m-2 kpjakb"
                tablesize="M"
              >
                <span
                  class="sc-1ntgl3l-0 ghpJwn"
                  textalign="initial"
                  textcolor="inherit"
                >
                  KRA
                </span>
              </button>
              <span
                class="sc-hjmm4m-1 cuEDpa"
              />
            </th>
            <th
              class="sc-hjmm4m-0 iOyNiA"
              tablesize="M"
            >
              <button
                align="center"
                class="sc-hjmm4m-2 bRawwG"
                tablesize="M"
              >
                <span
                  class="sc-1ntgl3l-0 ghpJwn"
                  textalign="initial"
                  textcolor="inherit"
                >
                  Weightage(%)
                </span>
              </button>
              <span
                class="sc-hjmm4m-1 cuEDpa"
              />
            </th>
          </tr>
        </thead>
        <tbody
          class="sc-c7cniy-0 kShdfA"
        >
          <tr
            class="sc-1yxujaa-0 bgyIwA"
            gridtemplatecolumns=" minmax(100px, 1fr) minmax(100px, 1fr) minmax(100px, 1fr)"
            style="grid-template-columns: minmax(100px, 1fr) minmax(100px, 1fr) minmax(100px, 1fr);"
            tabindex="0"
          >
            <td
              class="sc-1e51eaa-0 hFOqbY"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
            <td
              class="sc-1e51eaa-0 hFOqbY"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
            <td
              align="center"
              class="sc-1e51eaa-0 gwoPex"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
          </tr>
          <tr
            class="sc-1yxujaa-0 bgyIwA"
            gridtemplatecolumns=" minmax(100px, 1fr) minmax(100px, 1fr) minmax(100px, 1fr)"
            style="grid-template-columns: minmax(100px, 1fr) minmax(100px, 1fr) minmax(100px, 1fr);"
            tabindex="1"
          >
            <td
              class="sc-1e51eaa-0 hFOqbY"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
            <td
              class="sc-1e51eaa-0 hFOqbY"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
            <td
              align="center"
              class="sc-1e51eaa-0 gwoPex"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
          </tr>
          <tr
            class="sc-1yxujaa-0 bgyIwA"
            gridtemplatecolumns=" minmax(100px, 1fr) minmax(100px, 1fr) minmax(100px, 1fr)"
            style="grid-template-columns: minmax(100px, 1fr) minmax(100px, 1fr) minmax(100px, 1fr);"
            tabindex="2"
          >
            <td
              class="sc-1e51eaa-0 hFOqbY"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
            <td
              class="sc-1e51eaa-0 hFOqbY"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
            <td
              align="center"
              class="sc-1e51eaa-0 gwoPex"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
          </tr>
          <tr
            class="sc-1yxujaa-0 bgyIwA"
            gridtemplatecolumns=" minmax(100px, 1fr) minmax(100px, 1fr) minmax(100px, 1fr)"
            style="grid-template-columns: minmax(100px, 1fr) minmax(100px, 1fr) minmax(100px, 1fr);"
            tabindex="3"
          >
            <td
              class="sc-1e51eaa-0 hFOqbY"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
            <td
              class="sc-1e51eaa-0 hFOqbY"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
            <td
              align="center"
              class="sc-1e51eaa-0 gwoPex"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
          </tr>
          <tr
            class="sc-1yxujaa-0 bgyIwA"
            gridtemplatecolumns=" minmax(100px, 1fr) minmax(100px, 1fr) minmax(100px, 1fr)"
            style="grid-template-columns: minmax(100px, 1fr) minmax(100px, 1fr) minmax(100px, 1fr);"
            tabindex="4"
          >
            <td
              class="sc-1e51eaa-0 hFOqbY"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
            <td
              class="sc-1e51eaa-0 hFOqbY"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
            <td
              align="center"
              class="sc-1e51eaa-0 gwoPex"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </main>
</div>
`;
